File: lib/service/golist_client.dart
import 'dart:async';
import 'dart:convert';

import 'package:go_list/model/collections/shopping_list_collection.dart';
import 'package:go_list/model/shopping_list.dart';
import 'package:go_list/service/storage/provider/remote_storage_provider.dart';
import 'package:http/http.dart';
import 'package:web_socket_channel/web_socket_channel.dart';

class GoListClient {
  static const String backend =
      String.fromEnvironment("BACKEND", defaultValue: "192.168.178.58:8000");
  static const String originalBackend = String.fromEnvironment(
      "ORIGINAL_BACKEND",
      defaultValue: "192.168.178.58:8000");
  static const String environment =
      String.fromEnvironment("ENV", defaultValue: "dev");
  static const String apiKey =
      String.fromEnvironment("API_KEY", defaultValue: "123");

  WebSocketChannel? webSocketChannel;
  final Client client;
  String deviceId = "";

  GoListClient([Client? client]) : client = client ?? Client();

  String get httpProtocol => environment != "dev" ? "https" : "http";

  String get wsProtocol => environment != "dev" ? "wss" : "ws";

  Future<Response> _sendRequest(
      {required String endpoint,
      required HttpMethod httpMethod,
      Object? body}) async {
    String? jsonBody;
    if (body != null) {
      jsonBody = json.encode(body);
    }
    Uri uri = Uri.parse("$httpProtocol://$backend$endpoint");
    Response response;
    try {
      switch (httpMethod) {
        case HttpMethod.get:
          response = await client.get(uri, headers: headers);
          break;
        case HttpMethod.post:
          response = await client.post(uri, body: jsonBody, headers: headers);
          break;
        case HttpMethod.put:
          response = await client.put(uri, body: jsonBody, headers: headers);
        case HttpMethod.delete:
          response = await client.delete(uri, body: jsonBody, headers: headers);
      }
    } catch (e) {
      throw Exception("$httpMethod request to '$endpoint' failed: $e");
    }
    if ((response.statusCode ~/ 100) != 2) {
      throw Exception(
          "$httpMethod '$uri' failed, response code was ${response.statusCode}");
    }
    return response;
  }

  Future<String> createTokenToShareList(String shoppingListId) {
    return _sendRequest(
            endpoint: "/tokens",
            httpMethod: HttpMethod.post,
            body: {"shopping_list_id": shoppingListId})
        .then((response) => jsonDecode(utf8.decode(response.bodyBytes)))
        .then((responseJson) => responseJson["token"])
        // this must be the original backend url because only this one opens the app
        // the new backend url is only used as a temporary second api during the migration
        .then((token) => "$httpProtocol://$originalBackend/join?token=$token");
  }

  Future<ShoppingList> joinListWithToken(String token) {
    return _sendRequest(
            endpoint: "/tokens/join/$token", httpMethod: HttpMethod.post)
        .then((response) =>
            ShoppingList.fromJson(jsonDecode(utf8.decode(response.bodyBytes))));
  }

  Future<ShoppingList> getShoppingList(String shoppingListId) {
    return _sendRequest(
            endpoint: "/shopping-lists/$shoppingListId",
            httpMethod: HttpMethod.get)
        .then((response) =>
            ShoppingList.fromJson(jsonDecode(utf8.decode(response.bodyBytes))));
  }

  Future<ShoppingListCollection> getShoppingLists() {
    return _sendRequest(endpoint: "/shopping-lists", httpMethod: HttpMethod.get)
        .then((response) => ShoppingListCollection.fromJson(
            jsonDecode(utf8.decode(response.bodyBytes))));
  }

  Future<void> upsertShoppingList(ShoppingList shoppingList) async {
    await _sendRequest(
        endpoint: "/shopping-lists",
        httpMethod: HttpMethod.put,
        body: shoppingList.toJson());
  }

  Future<void> deleteShoppingList(String shoppingListId) async {
    await _sendRequest(
        endpoint: "/shopping-lists/$shoppingListId",
        httpMethod: HttpMethod.delete);
  }

  Map<String, String> get headers => {
        "api-key": apiKey,
        "user-id": deviceId,
        "Content-Type": "application/json"
      };

  Future<Stream<ShoppingList>> listenForChanges(String shoppingListId) async {
    // see https://github.com/dart-lang/http/issues/551
    Stream<ShoppingList>? stream =
        await runZonedGuarded<Future<Stream<ShoppingList>>>(() async {
      if (webSocketChannel != null) {
        webSocketChannel?.sink.close();
      }
      webSocketChannel = WebSocketChannel.connect(Uri.parse(
          "$wsProtocol://$backend/shopping-lists/$shoppingListId/listen"));
      webSocketChannel!.sink.add(json.encode(headers));
      return webSocketChannel!.stream.map((shoppingListJson) {
        return ShoppingList.fromJson(jsonDecode(shoppingListJson));
      });
    }, (error, stack) {
      throw Exception('Failed to connect to websocket');
    });
    if (stream == null) {
      throw Exception("Failed to connect to websocket");
    }
    return stream;
  }
}

File: svelte-golist/src/lib/services/GoListClient.ts
import { ShoppingList } from '../models/ShoppingList.js';
import { ShoppingListCollection } from '../models/collections/ShoppingListCollection.js';
import { HttpMethod } from '../types/index.js';

export class GoListClient {
  private static readonly BACKEND = import.meta.env.VITE_BACKEND || '192.168.178.58:8000';
  private static readonly ENVIRONMENT = import.meta.env.VITE_ENV || 'dev';
  private static readonly API_KEY = import.meta.env.VITE_API_KEY || '123';

  private webSocket: WebSocket | null = null;
  deviceId: string = '';

  get httpProtocol(): string {
    return GoListClient.ENVIRONMENT !== 'dev' ? 'https' : 'http';
  }

  get wsProtocol(): string {
    return GoListClient.ENVIRONMENT !== 'dev' ? 'wss' : 'ws';
  }

  private get headers(): Record<string, string> {
    return {
      'api-key': GoListClient.API_KEY,
      'user-id': this.deviceId,
      'Content-Type': 'application/json'
    };
  }

  private async sendRequest({
    endpoint,
    httpMethod,
    body
  }: {
    endpoint: string;
    httpMethod: HttpMethod;
    body?: any;
  }): Promise<Response> {
    const url = `${this.httpProtocol}://${GoListClient.BACKEND}${endpoint}`;
    let response: Response;
    
    const jsonBody = body ? JSON.stringify(body) : undefined;

    try {
      switch (httpMethod) {
        case HttpMethod.GET:
          response = await fetch(url, { method: 'GET', headers: this.headers });
          break;
        case HttpMethod.POST:
          response = await fetch(url, { method: 'POST', body: jsonBody, headers: this.headers });
          break;
        case HttpMethod.PUT:
          response = await fetch(url, { method: 'PUT', body: jsonBody, headers: this.headers });
          break;
        case HttpMethod.DELETE:
          response = await fetch(url, { method: 'DELETE', body: jsonBody, headers: this.headers });
          break;
        default:
          throw new Error(`Unsupported HTTP method: ${httpMethod}`);
      }
    } catch (error) {
      throw new Error(`${httpMethod} request to '${endpoint}' failed: ${error}`);
    }
    
    if (!response.ok) {
      throw new Error(`${httpMethod} '${url}' failed, response code was ${response.status}`);
    }
    
    return response;
  }

  async createTokenToShareList(shoppingListId: string): Promise<string> {
    const response = await this.sendRequest({
      endpoint: '/tokens',
      httpMethod: HttpMethod.POST,
      body: { shopping_list_id: shoppingListId }
    });
    
    const responseJson = await response.json();
    const token = responseJson.token;
    
    return `${this.httpProtocol}://${GoListClient.BACKEND}/join?token=${token}`;
  }

  async joinListWithToken(token: string): Promise<ShoppingList> {
    const response = await this.sendRequest({
      endpoint: `/tokens/join/${token}`,
      httpMethod: HttpMethod.POST
    });
    
    const json = await response.json();
    return ShoppingList.fromJson(json);
  }

  async getShoppingList(shoppingListId: string): Promise<ShoppingList> {
    const response = await this.sendRequest({
      endpoint: `/shopping-lists/${shoppingListId}`,
      httpMethod: HttpMethod.GET
    });
    
    const json = await response.json();
    return ShoppingList.fromJson(json);
  }

  async getShoppingLists(): Promise<ShoppingListCollection> {
    const response = await this.sendRequest({
      endpoint: '/shopping-lists',
      httpMethod: HttpMethod.GET
    });
    
    const json = await response.json();
    return ShoppingListCollection.fromJson(json);
  }

  async upsertShoppingList(shoppingList: ShoppingList): Promise<void> {
    await this.sendRequest({
      endpoint: '/shopping-lists',
      httpMethod: HttpMethod.PUT,
      body: shoppingList.toJson()
    });
  }

  async deleteShoppingList(shoppingListId: string): Promise<void> {
    await this.sendRequest({
      endpoint: `/shopping-lists/${shoppingListId}`,
      httpMethod: HttpMethod.DELETE
    });
  }

  async listenForChanges(shoppingListId: string): Promise<ReadableStream<ShoppingList>> {
    return new ReadableStream({
      start: (controller) => {
        if (this.webSocket) {
          this.webSocket.close();
        }

        const wsUrl = `${this.wsProtocol}://${GoListClient.BACKEND}/shopping-lists/${shoppingListId}/listen`;
        this.webSocket = new WebSocket(wsUrl);

        this.webSocket.onopen = () => {
          this.webSocket?.send(JSON.stringify(this.headers));
        };

        this.webSocket.onmessage = (event) => {
          try {
            const shoppingListJson = JSON.parse(event.data);
            const shoppingList = ShoppingList.fromJson(shoppingListJson);
            controller.enqueue(shoppingList);
          } catch (error) {
            console.error('Failed to parse websocket message:', error);
          }
        };

        this.webSocket.onerror = (error) => {
          console.error('WebSocket error:', error);
          controller.error(new Error('Failed to connect to websocket'));
        };

        this.webSocket.onclose = () => {
          controller.close();
        };
      },
      cancel: () => {
        this.webSocket?.close();
      }
    });
  }
}